packages/3dom/src/context-host.ts

```diff
@@ -0,0 +1,108 @@
+import {ThreeDOMCapability} from './api.js';
+import {ThreeDOMExecutionContext} from './context.js';
+import {ThreeDOMMessageType} from './protocol.js';
+
+// This fake worker class is just used to capture the transfer of the message
+// port.
+class FakeWorker extends EventTarget implements Worker {
+  onerror: EventListener|null = null;
+  onmessage: EventListener|null = null;
+  onmessageerror: EventListener|null = null;
+  port2: MessagePort|null = null;
+  hostExecutionContext: HostThreeDOMExecutionContext;
+
+  constructor(hostExecutionContext: HostThreeDOMExecutionContext) {
+    super();
+    this.hostExecutionContext = hostExecutionContext;
+  }
+
+  postMessage(message: any, transfer: Array<Transferable>): void;
+  postMessage(message: any, options?: PostMessageOptions|undefined): void;
+  postMessage(message: any, secondArg: any): void {
+    // When the handshake message is sent, capture and store the port.
+    if (message.type == ThreeDOMMessageType.HANDSHAKE) {
+      const transfer = secondArg as Array<Transferable>;
+      this.port2 = transfer[0] as MessagePort;
+      // Indicate the host execution context that the port has been captured.
+      this.hostExecutionContext.portHasBeenSet();
+    }
+  }
+
+  terminate() {
+    this.port2 = null;
+  }
+}
+
+const $iframe = Symbol('iframe');
+const $iframeLoaded = Symbol('iframeLoaded');
+
+export class HostThreeDOMExecutionContext extends ThreeDOMExecutionContext {
+  protected[$iframe]: HTMLIFrameElement;
+  protected[$iframeLoaded] = false;
+
+  constructor(
+      capabilities: Array<ThreeDOMCapability>,
+      iframe: HTMLIFrameElement|null = null) {
+    super(capabilities);
+    // Make sure there is an iframe to connect with
```

Comment:
this[$iframe] = iframe ? iframe : document.querySelector('iframe');
if (!this[$iframe]) {
  throw new Error(...
}